DiverseSelector/methods/similarity.py

```diff
@@ -0,0 +1,1138 @@
+        w_factor: {"fraction", "power_n"}
+            Type of weight function that will be used.
+            'fraction' : similarity = d[k]/n
+                            dissimilarity = 1 - (d[k] - n_objects % 2)/n_objects
+            'power_n' : similarity = n**-(n_objects - d[k])
+                        dissimilarity = n**-(d[k] - n_objects % 2)
+            other values : similarity = dissimilarity = 1
+
+        Returns
+        -------
+        counters : dict
+            Dictionary with the weighted and non-weighted counters.
+        """
+        # Check if the data is a np.ndarray of a list
+        if not isinstance(arr, np.ndarray):
+            raise TypeError(
+                "Input data is not a np.ndarray, to secure the right results please "
+                "input the right data type"
+            )
+        # Check if data is a columnwise sum or the objects
+        if arr.ndim == 1:
+            c_total = arr
+            # If data is a columnwise sum, check if n_objects is provided
+            if n_objects is None:
+                raise ValueError(
+                    "Input data is the columnwise sum, please specify number of objects"
+                )
+        else:
+            c_total = np.sum(arr, axis=0)
+            n_objects = len(arr)
+
+        # Assign c_threshold
+        if not c_threshold:
+            c_threshold = n_objects % 2
+        elif c_threshold == "dissimilar":
+            c_threshold = math.ceil(n_objects / 2)
+        elif isinstance(c_threshold, int):
+            if c_threshold >= n_objects:
+                raise ValueError(
+                    "c_threshold cannot be equal or greater than n_objects. \n"
+                    + f"c_threshold = {c_threshold}  n_objects = {n_objects}"
+                )
+            c_threshold = c_threshold
+        else:
+            raise ValueError(
+                "c_threshold must be None, 'dissimilar' or an integer. \n"
+                + f"Given c_threshold = {c_threshold}"
+            )
+
+        # Set w_factor function (a weight factor for the similarity and dissimilarity) is
+        # provided depending on the number of objects that are similar or not in a column
+        if w_factor:
+            # power_n case
+            if "power" in w_factor:
+                power = int(w_factor.split("_")[-1])
+
+                def f_s(d):
+                    return power ** -(n_objects - d).astype(float)
+
+                def f_d(d):
+                    return power ** -(d - n_objects % 2).astype(float)
+
+            # fraction case
+            elif w_factor == "fraction":
+
+                def f_s(d):
+                    return d / n_objects
+
+                def f_d(d):
+                    return 1 - (d - n_objects % 2) / n_objects
+
+        # default case, the similarity and dissimilarity counters are not weighted
+        else:
+
+            def f_s(d):
+                return 1
+
+            def f_d(d):
+                return 1
```

Comment:
@marco-2023, it can be better (and easier to understand) to add two private methods to the class, called `_f_s(d, w_factor)` and `_f_d(d, w_factor)` which return the desired value. Also, if `w_factor` is a class attribute (and stays fixed), it does not need to be passed as an argument. Then, they can be simply `staticmethods`.